### PR TITLE
Eval subdomain casting helper compatible with expression framework

### LINKED ISSUE #886: create helper functions to move between domain size
In https://github.com/o1-labs/proof-systems/blob/master/kimchi/src/circuits/domains.rs#L1 we define 4 domains of different sizes: d1, d2, d4, and d8. Each domain having twice as many elements are the previous one.

Evaluations in a specific domain size can quickly be casted in a _smaller_ domain by taking all the i-th elements, where i is the difference of size between the two domains. For example to go from d4 to d1, you'd pick every 4th element, and to go from d8 to d1 you'd pick every 8th element.

This results in a _lot_ of code that looks like this and is hard to parse:

```rust
let s = (plnm.domain().size / domain.size) as usize;
let v: Vec<_> = (0..(domain.size())).map(|i| plnm.evals[s * i]).collect();
```

sometimes it is even interlaced with some scaling:

```rust
x + evals.evals[(scale * i + (domain as usize) * shift) % evals.evals.len()]
```

it would be good to create helper functions to hide the low-level manipulation and make the code around it easier to parse. I can try to find more example of this pattern being used in the codebase, but grepping for these would be a good starting point!

### PR DESCRIPTION
**Pull request by jspada**
This PR builds on the awesome work of @duguorong009 and @mimoo to help get this improvement over the line.

Building on the PR #964 I found that in several places the expression framework does more than just add a constant, depending on the operation and situation.  It may
  * add or subtract a constant
  * multiply by a constant
  * merge two evals
  * add and subtract evaluation points
  * multiply evaluation points
  * square evaluation points
  
I also found that in many cases where domain casting is being done the circuit domain is not available.  Due to the above two issues, I extended the old API with a new function providing the required functionality.

Closes #886

### CODE REVIEW

File: kimchi/src/circuits/expr.rs [hunk 1]
```diff
@@ -1063,12 +1053,11 @@ impl<'a, F: FftField> EvalResult<'a, F> {
                     evals: es_sub,
                 },
             ) => {
-                let scale = (d_sub as usize) / (d as usize);
-                assert!(scale != 0);
-                evals.evals.par_iter_mut().enumerate().for_each(|(i, e)| {
-                    *e += es_sub.evals[(scale * i + (d_sub as usize) * s) % es_sub.evals.len()];
-                });
-                Evals { evals, domain: d }
+                evals =
```
Comment (mrmr1993) [inline kimchi/src/circuits/expr.rs:1056]: This (and other similar changes) change the memory allocation characteristics (`par_iter` vs `par_iter_mut`). Perhaps adding helpers that cache all the intermediate terms in a struct (`ReducedDomainEvals`?) and use them to index / iterate over the structure would help?

File: kimchi/src/circuits/expr.rs [hunk 2]
```diff
@@ -1082,23 +1071,14 @@ impl<'a, F: FftField> EvalResult<'a, F> {
...
-                        es1.evals[(scale1 * i + (d1 as usize) * s1) % es1.evals.len()]
-                            + es2.evals[(scale2 * i + (d2 as usize) * s2) % es2.evals.len()]
-                    })
-                    .collect();
-
                 Evals {
                     domain: res_domain.0,
-                    evals: Evaluations::<F, D<F>>::from_vec_and_domain(v, res_domain.1),
+                    evals: es1.to_subdomain_unsafe(res_domain.1, (d1 as usize) * s1, |eval, i| {
+                        eval + es2.evals[(scale2 * i + (d2 as usize) * s2) % es2.evals.len()]
```
Comment (mrmr1993) [inline kimchi/src/circuits/expr.rs:1080]: It's a pity that we still have to manually do this here.

Comment (jspada) [reply kimchi/src/circuits/expr.rs:1080]: Maybe there's a way to make a "merging" helper?

File: utils/src/evaluations.rs [hunk 3]
```diff
@@ -167,19 +174,53 @@ impl<F: FftField> ExtendedEvaluations<F> for Evaluations<F, Radix2EvaluationDoma
...
-        };
-        let new_evals = (0..target_size).into_par_iter().map(f).collect();
+        self.to_subdomain_unsafe(target_domain, shift, |eval, _| eval + cst)
+    }
+
+    fn to_subdomain_unsafe<G: Sync + Send + Fn(F, usize) -> F>(
+        &self,
+        target_domain: Radix2EvaluationDomain<F>,
+        shift: usize,
+        map: G,
```
Comment (mimoo) [inline utils/src/evaluations.rs:185]: imo passing a closure defeats the purpose of these helpers

File: utils/src/evaluations.rs [hunk 4]
```diff
@@ -167,19 +174,53 @@ impl<F: FftField> ExtendedEvaluations<F> for Evaluations<F, Radix2EvaluationDoma
...
+                .into_par_iter()
+                .map(|i| map(self.evals[(scale * i + shift) % self.evals.len()], i))
+                .collect(),
+            target_domain,
+        )
     }
 }
 
+/// Identity evaluation mapping
+pub const fn identity_fn<F: FftField>(eval: F, _: usize) -> F {
```
Comment (mimoo) [inline utils/src/evaluations.rs:220]: this should be inlined instead of being its own function

File: kimchi/src/circuits/expr.rs (hunk 2 above)
Comment (mimoo) [reply kimchi/src/circuits/expr.rs:1080]: yeah it'd be nicer to have a function that clearly states what it's doing instead

**Pull request closed without merging**